include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		} CASE_END
 #endif //NDF_BECKMANN
 
 #ifdef NDF_PHONG
-	if (ndf == NDF_PHONG) {
-		s = irr_glsl_beckmann_smith_cos_generate(currInteraction, rand.xy, ax, ay);
-	} else
+		CASE_BEGIN(ndf, NDF_PHONG) 
+		{
+			localH = irr_glsl_beckmann_cos_generate_wo_clamps(upperHemisphereLocalV, u.xy, ax, ay);
+		} CASE_END
+#endif //NDF_PHONG
+		CASE_OTHERWISE
+		{}
+		END_CASES
+
+		vec3 localL;
+		float VdotH = dot(localH, localV);
+		vec3 fr;
+		bool refraction = false;
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+		{
+			fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH);
+			rem *= fr;
+		}
+		else
+#endif
+		{
+			fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH);
+
+			const float refractionProb = colorToScalar(fr);
+			float rcpChoiceProb;
+			refraction = irr_glsl_partitionRandVariable(refractionProb, u.z, rcpChoiceProb);
+			pdf /= rcpChoiceProb;
+		}
+		float eta = colorToScalar(ior[0]);
+		float rcpEta = 1.0/eta;
+
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(refraction, localV, localH, localL, rcpEta, rcpEta*rcpEta);
+		s = irr_glsl_createLightSampleTangentSpace(localV, localL, tangentFrame);
+
+		const float TdotH2 = out_microfacet.TdotH * out_microfacet.TdotH;
+		const float BdotH2 = out_microfacet.BdotH * out_microfacet.BdotH;
+		const float NdotH2 = out_microfacet.isotropic.NdotH2;
+		const float NdotL = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+
+		BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+		CASE_BEGIN(ndf, NDF_GGX)
+		{
+			G2_over_G1 = irr_glsl_ggx_smith_G2_over_G1(NdotL, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, NdotV, TdotV2, BdotV2, NdotV2, ax2, ay2);
+			G1_over_2NdotV = irr_glsl_GGXSmith_G1_wo_numerator(NdotV, TdotV2, BdotV2, NdotV2, ax2, ay2);
+			ndf_val = irr_glsl_ggx_aniso(TdotH2, BdotH2, NdotH2, ax, ay, ax2, ay2);
+		} CASE_END
+#endif //NDF_GGX
+
+#ifdef NDF_BECKMANN
+		CASE_BEGIN(ndf, NDF_BECKMANN)
+		{
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, ax2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+		} CASE_END
+#endif //NDF_BECKMANN
+
+#ifdef NDF_PHONG
+		CASE_BEGIN(ndf, NDF_PHONG)
+		{
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL * s.TdotL, s.BdotL * s.BdotL, s.NdotL * s.NdotL, ax2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+		} CASE_END
 #endif //NDF_PHONG
-	{}
+		CASE_OTHERWISE
+		{}
+		END_CASES
+
+		rem *= G2_over_G1;
+		pdf *= irr_glsl_smith_VNDF_pdf_wo_clamps(ndf_val, G1_over_2NdotV);
```

Comment:
the PDF looks suspicious, you're using the VNDF to reflective differential factor transform here....

you should be using the one that can account for refraction
https://github.com/Devsh-Graphics-Programming/Nabla/blob/shader_pipeline_irrlichtbaw/include/irr/builtin/glsl/bxdf/geom/smith/common.glsl#L33